### PR TITLE
XNAT Integration

### PR DESCRIPTION
**Pull request by vferat**
- ADD support to save source data in the BIDS structure
- ADD overwrite option
- ADD support for` .mmf.zip,`  `.mff.tar`,` .mff.tar.gz` files.
- ADD basic test

These changes have not been integrated into the GUI. 
For now, it still works as before

I let you make the necessary changes if you want to integrate the new features to the GUI

### CODE REVIEW

File: egi2bids/mff2bids.py [hunk 1]
```diff
@@ -44,28 +50,71 @@
...
+
+
 # fmt: on
+def _extract_folder(file):
+    """Extracts a compressed folder to its original form"""
+    # Get the path and file extension
+    file = Path(file)
+    ext = file.suffix
+    # Open the archive file based on its type
+    print(file)
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
```

File: egi2bids/mff2bids.py [hunk 2]
```diff
@@ -44,28 +50,71 @@
...
     "Cz",
 ]
+
+
 # fmt: on
+def _extract_folder(file):
+    """Extracts a compressed folder to its original form"""
+    # Get the path and file extension
+    file = Path(file)
+    ext = file.suffix
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
    ext = "".join(file.suffixes)
    _check_value(ext, (".tar", ".tar.gz", ".zip", ".mff"), "extension")
```

more DRY + use `.suffixes` to handle file extensions.

File: egi2bids/mff2bids.py [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
+import numpy as np
 from mne_bids import (
     BIDSPath,
     make_dataset_description,
     update_sidecar_json,
     write_raw_bids,
 )
 
+from .utils._logs import logger
+
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
from .utils._logs import logger
from .utils._checks import _check_value
```

c.f. other comment.

File: egi2bids/mff2bids.py [hunk 4]
```diff
@@ -44,28 +50,71 @@
...
+        logger.info(f"Extracting zip archive to {file} to {path}")
+        archive = zipfile.ZipFile(file)
+        archive.extractall(path)
+        archive.close()
+
+    elif ext == ".mff":
+        path = file
+    else:
+        raise ValueError(f"Unsupported file type: {ext}")
+    print(path)
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
```

File: egi2bids/mff2bids.py [hunk 5]
```diff
@@ -44,28 +50,71 @@
...
 # fmt: on
+def _extract_folder(file):
+    """Extracts a compressed folder to its original form"""
+    # Get the path and file extension
+    file = Path(file)
+    ext = file.suffix
+    # Open the archive file based on its type
+    print(file)
+    if ext == ".tar":
+        path = file.with_name(file.name.replace(".tar", ""))
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
        path = file.with_name(file.name.replace(ext, ""))
```

I think this syntax is a bit safer as it relies on `ext = "".join(file.suffixes)` and reduces the amount of hard-coded variables.

File: egi2bids/mff2bids.py [hunk 6]
```diff
@@ -44,28 +50,71 @@
...
+        logger.info(f"Extracting zip archive to {file} to {path}")
+        archive = tarfile.open(file)
+        archive.extractall(path)
+        archive.close()
+
+    elif ext == ".gz":
+        path = file.with_name(file.name.replace(".tar.gz", ""))
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
    elif ext == ".tar.gz":
        path = file.with_name(file.name.replace(ext, ""))
```

File: egi2bids/mff2bids.py [hunk 7]
```diff
@@ -44,28 +50,71 @@
...
+        logger.info(f"Extracting zip archive to {file} to {path}")
+        archive = tarfile.open(file, "r:gz")
+        archive.extractall(path)
+        archive.close()
+
+    elif ext == ".zip":
+        path = file.with_name(file.name.replace(".zip", ""))
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
    elif ext == ".zip":
        path = file.with_name(file.name.replace(ext, ""))
```

File: egi2bids/mff2bids.py [hunk 8]
```diff
@@ -44,28 +50,71 @@
...
-    task=None,
-    session=None,
+    subject,
+    session,
+    task,
     run=None,
     event_id=None,
     line_frequency=50,
+    save_source=False,
+    overwrite=False,
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
    save_source=False,
    *,
    overwrite=False,
```

File: egi2bids/mff2bids.py [hunk 9]
```diff
@@ -44,28 +50,71 @@
...
 ):
-    bids_path = BIDSPath(root=Path(bids_root))
+    # mff_source
+    logger.info(mff_source)
+    # Extract
+    mff_source = _extract_folder(mff_source)
+
+    # BIDS path
+    bids_root = Path(bids_root)
+    bids_path = BIDSPath(root=bids_root)
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]: The update below can be done in a single line providing all arguments to the same call to `.update`, no?

File: egi2bids/mff2bids.py [hunk 10]
```diff
@@ -114,13 +171,12 @@ def mff2bids(
...
-        r"C:\Local\git\BIDS\datasets\study02",
-        subject="001",
-        task="MMN",
-        session="eeg",
-        run="002",
-    )
+    # save source
+    if save_source:
+        source_root = bids_root.joinpath("sourcedata")
+        print(f"saving source data to {source_root}")
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]:
```suggestion
        logger.info("Saving source data to %s", source_root)
```

File: egi2bids/tests/test_mff2bids.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+testing_path = data_path(download=True)
+egi_path = op.join(testing_path, "EGI")
+egi_mff_fname = op.join(egi_path, "test_egi.mff")
+
+
+@pytest.mark.parametrize("filetype", ["tar", "zip", "gztar"])
+def test_file_conversion(tmp_path, filetype):
+    # create temporary directory and file for testing
+    mff_path = op.join(tmp_path, "test.mff")
+    bids_path = op.join(tmp_path, "bids")
```
Comment (mscheltienne) [inline egi2bids/tests/test_mff2bids.py]:
```suggestion
    mff_path = tmp_path / "test.mff"
    bids_path = tmp_path / "bids"
```

File: egi2bids/tests/test_mff2bids.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        event_id=None,
+        line_frequency=50,
+        save_source=False,
+        overwrite=False,
+    )
+    assert BIDSLayout(bids_root, validate=True)
+
+
+def test_mff2bids_save_source(tmp_path):
+    bids_path = op.join(tmp_path, "bids")
```
Comment (mscheltienne) [inline egi2bids/tests/test_mff2bids.py]:
```suggestion
    bids_path = tmp_path / "bids"
```

File: egi2bids/commands/egi2bids.py [hunk 13]
```diff
@@ -12,35 +12,39 @@ def run():
     parser.add_argument(
         "mff_source",
         type=str,
-        help="path to the .mff file.",
+        help=(
+            "path to the input mff file. "
+            "Can be a '.mmf' folder or a '.mff.tar', "
```
Comment (mscheltienne) [inline egi2bids/commands/egi2bids.py]:
```suggestion
            "Can be a '.mff' folder or a '.mff.tar', "
```

File: egi2bids/commands/egi2bids.py [hunk 14]
```diff
@@ -12,35 +12,39 @@ def run():
...
         required=True,
     )
     parser.add_argument(
-        "-t",
-        "--task",
+        "-ses",
+        "--session",
         type=str,
         metavar="str",
-        help="Task.",
```
Comment (mscheltienne) [inline egi2bids/commands/egi2bids.py:35]: Shouldn't this be `type=int` and `metavar="int"`?

File: egi2bids/commands/egi2bids.py [hunk 15]
```diff
@@ -51,13 +55,29 @@ def run():
...
+        type=bool,
+        metavar="bool",
+        help="Either or not to save sourcedata.",
+        required=False,
+    )
+    parser.add_argument(
+        "--overwrite",
+        type=bool,
+        metavar="bool",
+        help="Either or not toallow overwritting data.",
```
Comment (mscheltienne) [inline egi2bids/commands/egi2bids.py]:
```suggestion
        help="Overwrites existing data.",
```

File: egi2bids/commands/egi2bids.py [hunk 16]
```diff
@@ -51,13 +55,29 @@ def run():
...
+        required=False,
+    )
```
Comment (mscheltienne) [inline egi2bids/commands/egi2bids.py]: Should have `action=store_true`, e.g.

```
parser.add_argument(
        "--my_bool_flag",
        help="if provided, stores True",
        action="store_true",
    )
```

**Review by mscheltienne — APPROVED** [review]
LGTM, I don't mind updating only the CLI part for now and GUI separately.
Feel free to disregard comments that you don't agree with, and to merge after.

Also:
- What about `.gz` only? (not `.tar.gz`)
- The `_extract_folder` repeats a lot of code and could be more DRY.

File: egi2bids/commands/egi2bids.py (hunk 14 above)
Comment (vferat) [reply egi2bids/commands/egi2bids.py:35]: from the BIDS specification [here](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#filesystem-structure) session is a `string`:
```If data for the subject were acquired across multiple sessions, then within the subject directory resides subdirectories named "ses-<label>", where string "<label>" is substituted with a unique identification label for each session.```

**Comment by vferat** [comment]
- I dropped the support for tar.gz for now.
- By default, Archive are now extracted in a temporary folder, which is deleted after program execution. 
- Add command argument to control log level
- mff2bids can now handle subfolder in `zip` and `tar `archives such as by searching the a `CONTENT` subfolder:
```
my_folder.mff.zip
│   ├── CONTENT
│   ├──  ...
│   ├──  workspace.plist
```

or 

```
my_folder.mff.zip
│   ├── my_folder.mff
│   │   ├──CONTENT
│   │   ├──...
│   │   ├── workspace.plist
```
TODO:
- delete temporary fiels from tests
- check what happens if BIDS does not overwrite and source_data overwrites

**Comment by vferat** [comment]
Tests are failing due to use of 
```
with (working_dir is not None else tempfile.TemporaryDirectory(suffix=".mff") as wd):
```

which is not handled by `python 3.8`

**Comment by mscheltienne** [comment]
You can bump to 3.9 or even 3.10 if you need. I don't mind dropping 3.8.

**Comment by vferat** [comment]
In the last batch of commits:

- Add fixture to generate temporary files during tests
- Check if `source_data `exists before writing `BIDS dataset` to prevent error in case of existing `source_data ` and non existing BIDS dataset)
- Remove support for python 3.8

Ready for review and merge

File: egi2bids/mff2bids.py [hunk 17]
```diff
@@ -47,58 +57,115 @@
...
+            archive.extractall(dir_)
+        for root, dirs, _ in os.walk(dir):
+            if "Contents" in dirs:
+                logger.info("MFF file found in %s", root)
+                return Path(root)
+        else:
+            raise (f"The '{ext}' archive does not contain a 'Content' folder.")
+
+    elif ext == ".mff":
+        return file
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]: @vferat I changed the structure of this function, with something a bit more DRY.

File: egi2bids/mff2bids.py [hunk 18]
```diff
@@ -47,58 +57,115 @@
...
+        # if event_id are not provided
+        if event_id is None:
+            event_ids = np.unique(events_data[:, 2])
+            event_id = {}
+            for event in event_ids:
+                event_id[f"Unkown_{event}"] = event
+        # TODO: check is provided events match events_data
+    else:
+        events_data = None
+        event_id = None
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py]: I did not check this part, but the hard-coding of STI 014 is a bit weird. It seems like a MEGIN channel, not like an EGI one (IIRC it's the combine channel on older triux systems, now called STI101 and STI102). 
In the MFF files I shared with you, what is the name of the trigger channel? (If there is one).

File: egi2bids/mff2bids.py [hunk 19]
```diff
@@ -47,80 +57,142 @@
...
+    task,
     run=None,
     event_id=None,
-    line_frequency=50,
+    save_source: bool = False,
+    working_dir=None,
+    *,
+    overwrite: bool = False,
+    verbose=None,
 ):
```
Comment (mscheltienne) [inline egi2bids/mff2bids.py:104]: I removed `line_frequency` as this is a CB-specific application, and is definitely not going to be used in the US ;)
Couple of questions on the arguments. The only one used in the CLI are:

mff_source, bids_root, subject, task, session, run, save_source, overwrite and verbose.

What is `working_dir` here for?

File: egi2bids/mff2bids.py [hunk 20]
```diff
@@ -47,80 +57,142 @@
...
 
-    # rename channels
-    new_chs = {}
-    for i, ch in enumerate(raw.info["ch_names"]):
-        if i > 256:
-            break
-        new_chs[ch] = ch_names_egi[i]
-    raw.rename_channels(new_chs)
+        # BIDS root
+        bids_root = _ensure_path(bids_root, must_exist=True)
```
Comment (vferat) [inline egi2bids/mff2bids.py]: ` must_exist=False` we should allow for the creation of a new BIDS dataset.

Comment (mscheltienne) [reply egi2bids/mff2bids.py]: Agree.

**Comment by mscheltienne** [comment]
@vferat I know I broke the test for now, sorry, will look into it; I got caught by something else this afternoon.

**Review by mscheltienne — APPROVED** [review]

**Comment by mscheltienne** [comment]
Good enough for now, thanks for fixing my typos.
To keep in mind, especially for the XNAT integration, we'll need to have a way to provide the events in the [CLI](https://github.com/fcbg-hnp-meeg/egi2bids/blob/main/egi2bids/commands/egi2bids.py)?

**Pull request merged**